[AArch64] use `isTRNMask` to calculate shuffle costs

ginsbach opened the pull request:
This builds on #169858 to fix the [divergence in codegen](https://godbolt.org/z/a9az3h6oq) between two very similar functions initially observed in #137447 (represented in the diff by test cases `@transpose_splat_constants` and `@transpose_constants_splat`:
```C++
int8x16_t f(int8_t x)
{
  return (int8x16_t) { x, 0, x, 1, x, 2, x, 3,
                       x, 4, x, 5, x, 6, x, 7 };
}

int8x16_t g(int8_t x)
{
  return (int8x16_t) { 0, x, 1, x, 2, x, 3, x,
                       4, x, 5, x, 6, x, 7, x };
}
```

The PR uses an additional `isTRNMask` call in `AArch64TTIImpl::getShuffleCost` to ensure that we treat shuffle masks as transpose masks even if `isTransposeMask` fails to recognise them (meaning that `Kind == TTI::SK_Transpose` cannot be relied upon).

Follow-up work could consider modifying `isTransposeMask`, but that would also impact other backends than AArch64.

ginsbach committed:
[AArch64] use isTRNMask to calculate shuffle costs

ginsbach committed:
[AArch64] add slp vectorizer test for new isTRNMask

davemgreen reviewed (commented):
I think that either the generic ShuffleVectorInst::isTransposeMask should match transposes like this, or the isTRNMask can be reused in the same places we currently check isZipMask/isUZPMask/etc.

Can you add direct cost-modelling tests for shuffles of these kinds?

ginsbach committed:
alternative implementation directly in getShuffleCost

ginsbach committed:
add cost model tests for trn with flipped inputs

ginsbach committed:
detect flipped trn miscategorised as SK_InsertSubvector

ginsbach commented:
@davemgreen I have added three more commits that should address your suggestions.

> I think that either the generic ShuffleVectorInst::isTransposeMask should match transposes like this, or the isTRNMask can be reused in the same places we currently check isZipMask/isUZPMask/etc.

With commit 3, I have picked the latter approach and used `isTRNMask` directly in `getShuffleCost` next to `isZipMask`, `isUZPMask`, etc. I have updated the PR description accordingly.

> Can you add direct cost-modelling tests for shuffles of these kinds?

Added in commit 4.

Two of the test cases added in commit 4 showed that things can go wrong with two-element shuffles
`shufflevector <2 x ???> %v0, <2 x ???> %v1, <2 x ???> <i32 2, i32 0>`.
The reason is that these get labelled as  `TTI::SK_InsertSubvector`. I have fixed the problem in commit 5.

MacDue committed:
Merge branch 'main' into isTRNMask-for-shuffle-cost

MacDue commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -6113,7 +6113,8 @@ AArch64TTIImpl::getShuffleCost(TTI::ShuffleKind Kind, VectorType *DstTy,
       LT.second.getVectorNumElements() == Mask.size() &&
-      (Kind == TTI::SK_PermuteTwoSrc || Kind == TTI::SK_PermuteSingleSrc) &&
+      (Kind == TTI::SK_PermuteTwoSrc || Kind == TTI::SK_PermuteSingleSrc ||
+       Kind == TTI::SK_InsertSubvector) &&
```
Maybe leave a comment about why `SK_InsertSubvector` is included here?

ginsbach replied:
Commit 7 adds such a comment.

MacDue reviewed (approved):
(Otherwise, LGTM)

ginsbach committed:
explanatory comment for SK_InsertSubvector

ginsbach commented:
> (Otherwise, LGTM)

Thank you. Please, would you merge this for me?
